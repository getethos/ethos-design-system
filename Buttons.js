import React from 'react'
import PropTypes from 'prop-types'

/* @getethos/design-system/Buttons.js

   Legend:

   - `Button` is a private component that returns an element with CSS classes.
   - `Button.scss` (not imported here) implements those CSS classes.
   - `ButtonFactory` is a HOC that creates public components with correct props.
   - `<Button.Tiny.Blue>`, etc. are the set of Design-approved public buttons.
   ========================================================================== */

/**
 * `Button` is a private component that returns a <button> with CSS classes and
 * other HTML attributes like `disabled` and `type`.
 *
 * Note that it is ignorant of which combinations Design considers legal or
 * illegal. Legal combinations must be declared via the enumerated exports at
 * the end of this file.
 *
 * The public props (Button.PUBLIC_PROPS) may be passed in downstream.
 * Other props may only be specified in this file.
 *
 * `type` is important to understand because it determines whether the button
 * is a regular button or a form submission button (<button type="submit">).
 * https://developer.mozilla.org/en/docs/Web/HTML/Element/button
 *
 * @param  {String}   props.children  The button text to display
 * @param  {Boolean}  props.disabled  Whether the button is disabled
 * @param  {String}   props.type      <button type="button|submit">
 * @param  {String}   props.size      The size of the button.
 * @param  {String}   props.style     The color style of the button.
 */
function Button({ children, disabled, type, size, style, ...unexpectedProps }) {
  const isValidHtmlType = Object.values(Button.HTML_TYPES).includes(type)
  const isValidSize = Object.values(Button.SIZES).includes(size)
  const isValidStyle = Object.values(Button.STYLES).includes(style)
  const unexpectedProp = Object.keys(unexpectedProps)[0]
  if (!isValidHtmlType) throw new TypeError(`Invalid HTML type '${type}'.`)
  if (!isValidSize) throw new TypeError(`Invalid size '${size}'.`)
  if (!isValidStyle) throw new TypeError(`Invalid style '${style}'.`)
  if (unexpectedProp) throw new TypeError(`Unexpected prop '${unexpectedProp}'`)

  return (
    <button
      className={['Button', size, style].join(' ')}
      disabled={disabled}
      type={type}
    >
      {children}
    </button>
  )
}

Button.HTML_TYPES = { BUTTON: 'button', SUBMIT: 'submit' } // read the docs^

Button.PUBLIC_PROPS = {
  children: PropTypes.string,
  disabled: PropTypes.bool,
  type: PropTypes.oneOf(Object.values(Button.HTML_TYPES)),
}

Button.SIZES = {
  LARGE: 'Large',
  MEDIUM: 'Medium',
  SMALL: 'Small',
  TINY: 'Tiny',
}

Button.STYLES = {
  BLUE: 'Blue',
  YELLOW: 'Yellow',
  WHITE: 'White',
  BLACK_OUTLINE: 'BlackOutline',
  BLUE_OUTLINE: 'BlueOutline',
  // TODO: StatefulButton, which is a distinct visual style plus nonstandard
  // functionality. The functionality will probably mean it's a new component,
  // and perhaps this file is just responsible for the markup/styles.
}

Button.propTypes = {
  ...Button.PUBLIC_PROPS,
  size: PropTypes.oneOf(Object.values(Button.SIZES)),
  style: PropTypes.oneOf(Object.values(Button.STYLES)),
}

Button.defaultProps = { type: Button.HTML_TYPES.BUTTON }

function ButtonFactory(privateProps) {
  function throwIllegalProp(prop) {
    const isIllegal = !Object.keys(Button.PUBLIC_PROPS).includes(prop)
    if (isIllegal) throw new TypeError(`Illegal prop '${prop}'`)
  }

  const PublicButtonComponent = (downstreamProps) => {
    Object.keys(downstreamProps).forEach(throwIllegalProp)
    return <Button {...downstreamProps} {...privateProps} />
  }

  return PublicButtonComponent
}

const PublicButtonComponents = {
  Large: {
    Blue: ButtonFactory({
      size: Button.SIZES.LARGE,
      style: Button.STYLES.BLUE,
    }),
    Yellow: ButtonFactory({
      size: Button.SIZES.LARGE,
      style: Button.STYLES.YELLOW,
    }),
    White: ButtonFactory({
      size: Button.SIZES.LARGE,
      style: Button.STYLES.WHITE,
    }),
    BlackOutline: ButtonFactory({
      size: Button.SIZES.LARGE,
      style: Button.STYLES.BLACK_OUTLINE,
    }),
    BlueOutline: ButtonFactory({
      size: Button.SIZES.LARGE,
      style: Button.STYLES.BLUE_OUTLINE,
    }),
  },

  Medium: {
    Blue: ButtonFactory({
      size: Button.SIZES.MEDIUM,
      style: Button.STYLES.BLUE,
    }),
    Yellow: ButtonFactory({
      size: Button.SIZES.MEDIUM,
      style: Button.STYLES.YELLOW,
    }),
    White: ButtonFactory({
      size: Button.SIZES.MEDIUM,
      style: Button.STYLES.WHITE,
    }),
    BlackOutline: ButtonFactory({
      size: Button.SIZES.MEDIUM,
      style: Button.STYLES.BLACK_OUTLINE,
    }),
    BlueOutline: ButtonFactory({
      size: Button.SIZES.MEDIUM,
      style: Button.STYLES.BLUE_OUTLINE,
    }),
  },

  Small: {
    Blue: ButtonFactory({
      size: Button.SIZES.SMALL,
      style: Button.STYLES.BLUE,
    }),
    Yellow: ButtonFactory({
      size: Button.SIZES.SMALL,
      style: Button.STYLES.YELLOW,
    }),
    White: ButtonFactory({
      size: Button.SIZES.SMALL,
      style: Button.STYLES.WHITE,
    }),
    BlackOutline: ButtonFactory({
      size: Button.SIZES.SMALL,
      style: Button.STYLES.BLACK_OUTLINE,
    }),
    BlueOutline: ButtonFactory({
      size: Button.SIZES.SMALL,
      style: Button.STYLES.BLUE_OUTLINE,
    }),
  },

  Tiny: {
    Blue: ButtonFactory({
      size: Button.SIZES.TINY,
      style: Button.STYLES.BLUE,
    }),
    Yellow: ButtonFactory({
      size: Button.SIZES.TINY,
      style: Button.STYLES.YELLOW,
    }),
    White: ButtonFactory({
      size: Button.SIZES.TINY,
      style: Button.STYLES.WHITE,
    }),
    BlackOutline: ButtonFactory({
      size: Button.SIZES.TINY,
      style: Button.STYLES.BLACK_OUTLINE,
    }),
    BlueOutline: ButtonFactory({
      size: Button.SIZES.TINY,
      style: Button.STYLES.BLUE_OUTLINE,
    }),
  },
}

export { PublicButtonComponents as Button }
