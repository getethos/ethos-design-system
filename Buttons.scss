/* @getethos/design-system/Buttons.scss
   ========================================================================== */

// Defaults

button,
.Button {
  font-family: var(--UntitledSans-font-stack);
  font-weight: 500;
  line-height: 1.25;
  border-radius: 3px;
  border: 1px solid transparent;

  transition: var(--transition-default);
  transition-property: background-color, border-color, color;

  &:not([disabled]) {
    cursor: pointer;
  }
}

// Sizes

.Button.Large {
  font-size: var(--L-font-size);
  padding: 16px 23px;
}

.Button.Medium {
  font-size: var(--M-font-size);
  padding: 14px 23px;
}

.Button.Small {
  font-size: var(--S-font-size);
  padding: 10px 15px 11px;
}

.Button.Tiny {
  font-size: var(--S-font-size);
  padding: 6px 11px 7px;
}

.Button.FullWidth {
  // This could modify any of the above sizes.
  width: 100%;
}

// Styles

@mixin light-on-dark-font-smoothing {
  -webkit-font-smoothing: antialiased; // stackoverflow.com/a/17927764
  -moz-osx-font-smoothing: grayscale;
}

.Button.Blue {
  background-color: var(--BrandBlue--default);
  color: var(--White);
  @include light-on-dark-font-smoothing;

  &:not([disabled]) {
    &:hover {
      background-color: var(--BrandBlue--hover);
    }

    &:active {
      background-color: var(--BrandBlue--active);
    }
  }

  &[disabled] {
    background-color: var(--GrayLightest--translucent);
    color: var(--White);
  }
}

.Button.Yellow {
  background-color: var(--BrandYellow--default);
  color: var(--Black--opaque);

  &:not([disabled]) {
    &:hover {
      background-color: var(--BrandYellow--hover);
    }

    &:active {
      background-color: var(--BrandYellow--active);
    }
  }

  &[disabled] {
    background-color: var(--GrayLightest--translucent);
    color: var(--White);
  }
}

.Button.White {
  background-color: var(--White);
  color: var(--Black--opaque);
  box-shadow: 0px 1px 1px rgba(0, 0, 0, 0.1), 0px 1px 4px rgba(0, 0, 0, 0.1);

  &:not([disabled]) {
    &:hover {
      color: var(--BrandBlue--hover);
    }

    &:active {
      color: var(--BrandBlue--active);
    }
  }

  &[disabled] {
    background-color: var(--GrayLightest--translucent);
    color: var(--White);
    @include light-on-dark-font-smoothing;
    box-shadow: none;
  }
}

.Button.BlackOutline {
  border-color: var(--Black--opaque);
  background-color: var(--White--transparent);
  color: var(--Black--opaque);

  &:not([disabled]) {
    &:hover {
      background-color: var(--Black--opaque);
      color: var(--White);
    }

    &:active {
      background-color: var(--Black--opaque);
      color: var(--White);
    }
  }

  &[disabled] {
    border-color: var(--GrayLightest--translucent);
    color: var(--GrayLightest--translucent);
  }
}

.Button.BlueOutline {
  border-color: var(--BrandBlue--default);
  background-color: var(--White--transparent);
  color: var(--BrandBlue--default);

  &:not([disabled]) {
    &:hover {
      background-color: var(--BrandBlue--hover);
      color: var(--White);
    }

    &:active {
      background-color: var(--BrandBlue--active);
      color: var(--White);
    }
  }

  &[disabled] {
    border-color: var(--GrayLightest--translucent);
    color: var(--GrayLightest--translucent);
  }
}
