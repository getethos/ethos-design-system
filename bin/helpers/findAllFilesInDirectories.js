const findAllFilesWithExt = require('./findAllFilesWithExt')

module.exports = function findAllFilesInDirectory(fileDirectories, extension) {
  return fileDirectories.reduce((memo, directory) => {
    return memo.concat(findAllFilesWithExt(directory, extension))
  }, [])
}
