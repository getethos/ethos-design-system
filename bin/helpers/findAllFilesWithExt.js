const fs = require('fs')
const path = require('path')

const findFilesWithExt = (directoryPath, extension) => {
  if (!fs.existsSync(directoryPath)) {
    console.log('No directory found: ', directoryPath)
    return []
  }

  const fileNames = fs.readdirSync(directoryPath)
  if (!fileNames.length) {
    return []
  }

  return fileNames.reduce((accumulatedFilePaths, fileName) => {
    const filePath = path.join(directoryPath, fileName)
    if (filePath.includes('node_modules')) return accumulatedFilePaths

    const stat = fs.lstatSync(filePath)
    const ext = path.extname(filePath)
    if (stat.isDirectory()) {
      return accumulatedFilePaths.concat(findFilesWithExt(filePath, extension))
    } else if (ext === extension) {
      return accumulatedFilePaths.concat(filePath)
    }
    return accumulatedFilePaths
  }, [])
}

module.exports = findFilesWithExt
