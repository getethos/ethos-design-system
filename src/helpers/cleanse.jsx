// first replace removes text-mask's underscores, second fixes like: '2//'
export default (value) => value.replace(/[,_$]/g, '').replace('//', '/')
