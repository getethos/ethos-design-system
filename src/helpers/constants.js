// https://airbnb.io/javascript/#naming--uppercase

export const INIT_INVALID = 'INVALID'

export const codes = {
  PAGE_UP: 33,
  PAGE_DOWN: 34,
  ESCAPE: 27,
  RETURN: 13,
  SPACE: 32,
  END: 35,
  HOME: 36,
  LEFT: 37,
  UP: 38,
  RIGHT: 39,
  DOWN: 40,
  TAB: 9,
}
