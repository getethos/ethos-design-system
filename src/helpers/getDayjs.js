// https://github.com/iamkun/dayjs/issues/480
import * as dayjs_ from 'dayjs'
const dayjs = dayjs_.default || dayjs_

import customParseFormat from 'dayjs/plugin/customParseFormat'
dayjs.extend(customParseFormat)

// Adds the UTC plugin
import dayjsPluginUTC from 'dayjs-plugin-utc'
dayjs.extend(dayjsPluginUTC)

export default dayjs
