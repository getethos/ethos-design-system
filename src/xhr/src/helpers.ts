import XhrError from './XhrError'

export interface KeyValMap {
  [key: string]: any
}

export interface XhrComponentState {
  status?: string
  error?: XhrError
  successMessage?: string
}

export type FormEvent = React.FormEvent<HTMLFormElement>

