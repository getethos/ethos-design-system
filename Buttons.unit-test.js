import React from 'react'

import { deepSnapshot } from 'test/helpers/snapshot'

import { Button } from './Buttons'

const buttonText = 'Click me'

describe('Buttons.js', () => {
  describe('{ Button }', () => {
    Object.keys(Button).forEach((Size) => {
      Object.keys(Button[Size]).forEach((Style) => {
        describe(`<Button.${Size}.${Style}>`, () => {
          it('matches snapshot', () => {
            const Component = Button[Size][Style]
            expect(
              deepSnapshot(<Component>{buttonText}</Component>)
            ).toMatchSnapshot()
          })
        })
      })
    })
  })
})
