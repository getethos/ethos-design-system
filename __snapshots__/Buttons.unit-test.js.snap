// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons.js { Button } <Button.Medium.Black> matches snapshot 1`] = `
<button
  className="Button Medium Black"
  type="button"
>
  Click me
</button>
`;

exports[`Buttons.js { Button } <Button.Medium.BlackOutline> matches snapshot 1`] = `
<button
  className="Button Medium BlackOutline"
  type="button"
>
  Click me
</button>
`;

exports[`Buttons.js { Button } <Button.Medium.Stateful> matches snapshot 1`] = `
<button
  className="Button Medium Stateful"
  type="button"
>
  Click me
</button>
`;

exports[`Buttons.js { Button } <Button.Medium.WhiteOutline> matches snapshot 1`] = `
<button
  className="Button Medium WhiteOutline"
  type="button"
>
  Click me
</button>
`;
